synchro-spec/ffg.tla

```diff
@@ -0,0 +1,206 @@
+
+\* @type: ($ffgVote) => Bool;
+IsValidFFGVote(vote) ==
+    /\ IsValidCheckpoint(vote.source)
+    /\ IsValidCheckpoint(vote.target)
+    /\ Edge(vote.target[1], vote.source[1]) \in block_graph_closure
+    /\ vote.source[2] < vote.target[2] 
+
+\* @type: ($checkpoint, Set($vote), Set($checkpoint)) => Bool;
+IsJustified(checkpoint, viewVotes, fixpoint) == 
+    \/ checkpoint = GenesisCheckpoint
+    \/ LET validatorsWhoCastJustifyingVote == { 
+        v \in VALIDATORS: \E justifyingVote \in viewVotes:
+            /\ justifyingVote.validator = v
+            /\ LET ffgVote == justifyingVote.ffg_vote IN
+                \* L6:
+                /\ ffgVote.source \in fixpoint
+                \* L7:
+                /\ <<ffgVote.target[1], checkpoint[1]>> \in block_graph_closure
+                /\ <<checkpoint[1], ffgVote.source[1]>> \in block_graph_closure
+                \* L8:
+                /\ ffgVote.target[2] = checkpoint[2] }    
+        IN 3 * Cardinality(validatorsWhoCastJustifyingVote) >= 2 * N
+
+\* @type: ($checkpoint, Set($vote), Set($checkpoint)) => Bool;
+IsFinalized(checkpoint, viewVotes, justifiedCheckpoints) ==
+    /\ checkpoint \in justifiedCheckpoints
+    /\ LET validatorsWhoCastFinalizingVote == { 
+        v \in VALIDATORS: \E finalizingVote \in viewVotes:
+            /\ finalizingVote.validator = v
+            /\ LET ffgVote == finalizingVote.ffg_vote IN
+                \* L14:
+                /\ ffgVote.source = checkpoint
+                \* L15:
+                /\ ffgVote.target[2] = checkpoint[2] + 1 }
+        IN 3 * Cardinality(validatorsWhoCastFinalizingVote) >= 2 * N
+
+AreConflictingBlocks(b1, b2) ==
+    /\ Edge(b1,b2) \notin block_graph_closure
+    /\ Edge(b2,b1) \notin block_graph_closure
+
+
+Vote(validator, ffgVote) == [
+    validator |-> validator,
+    ffg_vote |-> ffgVote
+]
+
+(*
+The most notable difference in the synchronous spec is that 
+If, at current_slot, a checkpoint with a slot number i < current_slot
+has not been justified, there will never be a vote justifying it in the future,
+because the votes are always cast s.t. the tharget checkpoint slot is current_slot
+
+This allows us to compute justified_checkpoints incrementally.
+*)
+\* @type: ($checkpoint, $checkpoint, Set(Str)) => Bool;
+CastVotes(source, target, validators) ==
+    LET ffgVote == [ source |-> source, target |-> target ] IN
+    /\ UNCHANGED <<blocks, block_graph, block_graph_closure>>
+    /\ IsValidFFGVote(ffgVote)
+    /\ validators /= {}
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    \* The ffg vote (b1, s1) -> (b2, s2) can justify (b,s2), for all blocks b1 <= b <= b2 in the block graph
+    \* We branch, based on whether there is a quorum of the ffg vote
+    /\ 
+        IF (3 * Cardinality({v \in VALIDATORS: Vote(v, ffgVote) \in votes'}) >= 2 * N /\ source \in justified_checkpoints)
```

Comment:
Is it really enough to check for the specific `ffgVote` here?
Couldn't there be other (longer or shorter) votes with target `target` that become justifying because of the votes cast in this action?

Reply:
I think you're right, I misunderstood that in the validator quorum for one particular checkpoint, each validator could have technically cast a different ffg vote, they don't all have to have the same one `k`-times.